### ISSUE TITLE
arm64_v8a: ibeta_inv exceeds 200 iterations for some arguments

### ISSUE THREAD

**Issue by linolinco** [boostorg/math#442]
Hi, when compiling for 64-bit ARM (arm64_v8a), ibeta_inv(alpha, beta, x) produces

boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::math::evaluation_error> >: Error in function boost::math::ibeta<long double>(long double, long double, long double): Root finding evaluation exceeded 200 iterations, giving up now.

when either alpha or beta are smaller than about 3e-3. For other target processors I've tried (arm_v7a, x86, x86_64) this works fine.

I've now hacked up my own custom root finder for these cases which converges in just a few iterations. My code's certainly not up to boost standards, but I'm happy to share it as a starting point if someone wants to use it to improve the ibeta_inv root finder in this regard.

Though I suspect that something else is going on here - the bug may be indicative of some deeper numerical issues on arm64_v8a?

**Comment by jzmaddock** [comment]
Works for me here, but may possibly be fixed by https://github.com/boostorg/math/commit/fa8f763df9e6635cf255d60766af65950eec58e8.

Can you:

1) check to see if the above commit fixes your issue, otherwise,
2) Can you please provide a test case (the exact arguments that trigger the issue), and
3) Compile your test case with `-DBOOST_MATH_INSTRUMENT` and then run your test case and report the program output - it'll allow me to see where the root finder is going wrong.

Many thanks for the report!

**Comment by linolinco** [comment]
Hi, fa8f763 unfortunately didn't help. Here is the test case I've boiled it down to:
[test.cpp.txt](https://github.com/boostorg/math/files/5329872/test.cpp.txt)
[test.pro.txt](https://github.com/boostorg/math/files/5329875/test.pro.txt)
(remove .txt extensions)

I'm using Qt 5.15.1 as IDE so I'm afraid there's still some baggage along for the ride, though there's no actual dependency on Qt other than its toolchain. FWIW Qt's compilation for this is:

/Developer/Applications/AndroidSDK/ndk/21.1.6352462/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ -c -target aarch64-linux-android21 -fno-limit-debug-info -fPIC -fstack-protector-strong -DANDROID -g -Wall -W -D_REENTRANT -fPIC -DBOOST_MATH_INSTRUMENT -DQT_QML_DEBUG -I../tools -I. -I/Developer/Qt/5.15.1/android/mkspecs/android-clang -o arm64-v8a/test.o ../tools/test.cpp

/Developer/Applications/AndroidSDK/ndk/21.1.6352462/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ -target aarch64-linux-android21 -fno-limit-debug-info -Wl,-soname,libtest_arm64-v8a.so -Wl,--build-id=sha1 -Wl,--no-undefined -Wl,-z,noexecstack -shared -o libtest_arm64-v8a.so arm64-v8a/test.o      -llog -lz -lm -ldl -lc

in case you can glean anything from that. BOOST_MATH_INSTRUMENT unfortunately doesn't help  as I can't see cout/cerr under Qt.

Can anyone help us run this naked (without Qt) on an arm64_v8a device?

**Comment by jzmaddock** [comment]
Well.... I've just spent 2 rather fruitless days getting Android Studio up and running (how does anyone actually get any work done with this thing?) only to resolutely fail to be able to run any of the arm virtual machines :(  The Intel ones are fine (albeit slow as hell), just can't get an arm emulation one to start up.

So.... I'm going to have to get you to do some debugging at your end.

What's happening in this domain is that the initial guess is zero (actually gets set to `boost::math::tools::min_value<T>()`), the derivatives are very close to infinity, and so after some checks and mucking about, we should go straight to the result (`0`) with no iterations at all.

If you define BOOST_MATH_INSTRUMENT when building and then do something like this:

```
   std::stringstream ss;
   auto p = std::cout.rdbuf(ss.rdbuf());

   std::cout << boost::math::ibeta_inv(1e-3, 20.0, 0.2) << std::endl;

   std::cout.rdbuf(p);
   // do something with ss.str() to get the output here...
```

We might hopefully get some useful information.  My guess is that something somewhere is ending up as a NaN, though I don't see how/why.

**Comment by linolinco** [comment]
I feel your pain. Qt helps insulate me somewhat from the hell of mobile development, though not nearly enough. (This also explains my interest in WebAssembly.) Yes the Android emulator only emulates the actual CPU it runs on... so to debug ARM code you need to physically hook up an ARM device via USB, very 20th century. Well at least my smartphone charges while it serves.

Okay let's see. First, I verified that all instances where the iteration limit throws in my code are indeed ones where the result should be zero. Second, here is the output from the instrumented ibeta_inv(1e-3, 20.0, 0.2) call:

Second order root iteration, guess = 3.3621031431120935062626778173217526e-4932, min = 3.3621031431120935062626778173217526e-4932, max = 1, digits = 84, max_iter = 200
Second order root iteration, limit = 1.0339757656912845935892608650874536e-25
boost/math/special_functions/beta.hpp:999 a = 0.0010000000000000000208166817117216851
boost/math/special_functions/beta.hpp:1000 b = 20
boost/math/special_functions/beta.hpp:1001 x = 3.3621031431120935062626778173217526e-4932
boost/math/special_functions/beta.hpp:1002 inv = 0
boost/math/special_functions/beta.hpp:1003 normalised = 1
boost/math/special_functions/beta.hpp:1188 fract = 0
boost/math/tools/roots.hpp:563 f0 = -0.2000000000000000111022302462515654
boost/math/tools/roots.hpp:564 f1 = 2.1517460115917398440081138030859217e-4930
boost/math/tools/roots.hpp:565 f2 = 0
Second order root iteration, delta = -9.2947773074783736806962835155767105e+4928
boost/math/tools/roots.hpp:<exception triggered>

Anything informative?

**Comment by jzmaddock** [comment]
It might have been more useful to see the output from the first iteration, but whatever those derivative values (f1 and f2) are completely wrong :(

Quick sanity check - what happens if you build with -DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false ?

That will stop the code from using long doubles internally, I'm guessing there's a bug in the std lib long double support code somewhere, but tracking it down might be a bind!

**Comment by linolinco** [comment]
What I wrote was the entire output.

With -DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false the call works and returns zero as expected.

Happy to help - but since you've already gone through the pain of setting up Android Studio, if you have an Android phone lying around, might as well hook it up?

**Comment by jzmaddock** [comment]
>What I wrote was the entire output - should there have been more?

Yes that was only the final iteration of the root finder's main loop.

>With -DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false the call works and returns zero as expected.

I suspect that will have to become the "official" workaround.

>Happy to help - but since you've already gone through the pain of setting up Android Studio, if you have an Android phone lying around, might as well hook it up?

Sorry, no andriod devices around here :(

>FWIW, my workaround for the time being is:

It would be better to disable long double usage as noted above.

BTW we have a long double test program to probe issues like this: https://github.com/boostorg/math/blob/develop/test/test_long_double_support.cpp.  However, figuring out how to run it on a mobile device might be a challenge!

Can you also see if `-DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false` fixes the web assembly issue?

**Comment by linolinco** [comment]
Yes, -DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false also fixes the WebAssembly issue. Thanks!

If it's straight C++ running the test program on arm64_v8a should be no problem, that's what Qt's for. I'll let you know.
